quantum/debounce/sym_defer_g.c

```diff
@@ -24,23 +24,31 @@ When no state changes have occured for DEBOUNCE milliseconds, we push the state.
 #    define DEBOUNCE 5
 #endif
 
+// Maximum debounce: 255ms
+#if DEBOUNCE > UINT8_MAX
+#    undef DEBOUNCE
+#    define DEBOUNCE UINT8_MAX
+#endif
+
 #if DEBOUNCE > 0
 static bool         debouncing = false;
 static fast_timer_t debouncing_time;
 
 void debounce_init(uint8_t num_rows) {}
 
 bool debounce(matrix_row_t raw[], matrix_row_t cooked[], uint8_t num_rows, bool changed) {
-    bool cooked_changed = false;
-
     if (changed) {
         debouncing      = true;
         debouncing_time = timer_read_fast();
+        return false;
```

Comment:
This `return false` statement will add quite a number of bytes for AVR and it is the _default_ debounce algorithm. I'm curious how it will prevernt sym_defer_g from possibly skipping debounce.

Reply:
The logic seems to be that if the debounce algorithm has just set the `debouncing_time` variable, it can't be time to actually debounce yet, especially since the only information available to debounce is the matrix contents in `raw` and `cooked` from the same invocation of `debounce`, and we need to wait for the next one. In that sense, this is a CPU time optimization.

Applying just the changes in this function to sofle:default, which uses `sym_defer_g` on AVR, causes an increase of 2 bytes.

Reply:
Exactly. I can move things around to try and make it smaller again. Also, combined with the changes to the TIME_DIFF macros, what is the total diff in size? I can't check right now. I got rid of a lot of branching there.

Reply:
Thanks for the clarification, so it is not feature change but cycle optimisation. Its impact approximately 8 bytes on other boards, such as `ferris/sweep`. I presume the _optimisation_ is just one single tick / cycle?

Reply:
Basically, I don't want to call the timer a second time and there's nothing that guarantees that we won't fall on some weird corner case scenario where the first timer call to set the `debouncing_time` will return exactly the same value as the second one, and that we won't accidentally skip debouncing when e.g. DEBOUNCE is set to 1.

Reply:
Applying the entire PR onto current `develop` is also a 2 byte increase on sofle:default (with `sym_defer_g`). So there doesn't seem to be something else optimizing the size in the rest of the code to balance out the increase. The branching in `TIMER_DIFF_*` already got optimized away because the elements of the computation that determines the requested bit count are compile-time constants.

Reply:
> [@filterpaper] Thanks for the clarification, so it is not feature change but cycle optimisation. Its impact approximately 8 bytes on other boards, such as `ferris/sweep`. I presume the _optimisation_ is just one single tick / cycle?

The code that gets skipped is as follows:
```c
    if (debouncing && timer_elapsed_fast(debouncing_time) >= DEBOUNCE) {
        if (memcmp(cooked, raw, sizeof(matrix_row_t) * num_rows) != 0) {
            memcpy(cooked, raw, sizeof(matrix_row_t) * num_rows);
            cooked_changed = true;
        }
        debouncing = false;
    }
```
Depending on the matrix's row count, avoiding `memcmp` and `memcpy` can be big.

Reply:
I'll see if I can manage to:

1. revert some of the changes/get rid of memset
2. make this smaller by replacing the separate return with an "else if" and keeping a single return statement

Either way, unless we can *always* guarantee in *all* architectures that two subsequent calls to timers will get the same value within this function (and I guess we can't, because ARM, for example (I'm almost certain?), can have the timer pass maybe 1ms in this time (between the latest nanos from one ms to the first nanos of the next, considering it's all integers)... we need to change this logic in order to ensure it absolutely never skips debounce.

Reply:
> The code that gets skipped is as follows:

"`debouncing_time = timer_read_fast();` should cause `timer_elapsed_fast(debouncing_time) >= DEBOUNCE` to evaluate `false` within the same _tick_?

Reply:
> > The code that gets skipped is as follows:
> 
> "`debouncing_time = timer_read_fast();`" should cause "`timer_elapsed_fast(debouncing_time) >= DEBOUNCE`" to evaluate `false` within the same _tick_?

Guaranteed to be the case in all platforms, including future ones?

I'd rather get the time once and use it twice with TIME_DIFF than to call thing sort of thing twice.

Reply:
> Guaranteed to be the case in all platforms, including future ones?

That was my original curiosity and technical interest — how or what platform will it be not guaranteed?

Reply:
An interrupt could also happen between `debouncing_time = ...;` and `timer_elapsed_fast(debouncing_time)` in the same invocation of `debounce`, which would trigger the debounce to happen right then and there (with the same data, not a fresh matrix report) if the interrupt took `DEBOUNCE` ms.

There is an argument to be made for `if (changed) debouncing = true; else ...` being **more correct** than the current code. We **need** to get out as soon as possible so that a new matrix report can be given to a second invocatien of `debounce`.

In my opinion, that is worth making - and *thoroughly testing* - a second PR for, in all debounce implementations.

Reply:
Even if we currently have no platforms in which that could happen, the algorithm itself shouldn't be susceptible to such scenario. If we ever port QMK to a platform in which this is possible, it completely breaks. I think ARM could even be such a case right now, since the clock is derived from CPU cycles? Please correct me if I'm wrong.

Regarding the other algorithms, @Nebuleon, at least the `_pk` ones aren't affected, _I think_. First and foremost because they only invoke the timer once and then calculate the difference using `TIME_DIFF_FAST`, so no multiple invocations would take place, and since the timer is per key, it can't trigger a premature report. Not from what I checked, that is. I didn't check much about `_pr` though, but I did make some changes to one of them.

Reply:
> Applying just the changes in this function to sofle:default, which uses `sym_defer_g` on AVR, causes an increase of 2 bytes.

It will be worth pointing out that AVR boards close to size limit have LTO enabled. The byte difference with link time optimisation seems large large and non-trivial:

```c
qmk compile -c -kb cradio -km default -e LTO_ENABLE=yes
 * The firmware size is fine - 16742/28672 (58%, 11930 bytes free)
```
With the `return false` statement:
```c
qmk compile -c -kb cradio -km default -e LTO_ENABLE=yes
 * The firmware size is fine - 16790/28672 (58%, 11882 bytes free)
```

Reply:
Eep, that's a lot of bytes indeed...

Reply:
The size increase does not impact all equally. There might be something else influencing the change.

Reply:
Running 

```bash
qmk compile -c -kb cradio -km default -e LTO_ENABLE=yes
```

using arm-none-eabi-gcc.exe (GCC) 12.2.0, I get:

* Original code:
```
* The firmware size is fine - 16692/28672 (58%, 11980 bytes free)
```

* With `const size_t matrix_size`:
```
 * The firmware size is fine - 16692/28672 (58%, 11980 bytes free)
 ```

* With an 'else if' + `const size_t matrix_size`:
```
* The firmware size is fine - 16700/28672 (58%, 11972 bytes free)
```

* With a `return false` + `const size_t matrix_size`:
```
* The firmware size is fine - 16700/28672 (58%, 11972 bytes free)
```

Reply:
Using what version of `avr-gcc`? `cradio` uses an AVR atmega32u4 CPU.

Reply:
```
avr-gcc.exe (GCC) 12.2.0
Copyright (C) 2022 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

I think it's whatever is right now on MSYS2 (I use QMK MSYS and I ran `pacman -Syuu` before these tests).

Reply:
`avr-gcc` 8.5 is recommended version installed by the script and used by the CI image because it produces the smallest binary.

That said, this looks like a compiler optimisation problem and the CI build passed for everything; so it should be immaterial for this PR.

Reply:
> `avr-gcc` 8.5 is the recommended version installed by the script and used by the CI image because it produces the smallest binary.
> 
> That said, this looks like a compiler optimisation problem and the CI build passed for everything; so it should be immaterial for this PR.

But my binaries with avr-gcc 12.2.0 are smaller than the ones generated by CI, aren't they? What version did you test that with?

Reply:
Your default binaries, without changes from this PR, is larger than the ones built by v8.5. That's the version I used for tests.

Reply:
Ok, I gotta check it again with 8.5, because with 12.2.0, I got 16692 bytes without the return, while you got 16742. And I got 16700 with the return, whereas you said you got 16790.

Either those numbers are wrong or 12.2.0 is generating smaller binaries.

Reply:
My apologies, I was reading them incorrectly (before coffee) as firmware output size, instead of free bytes. I think the smaller size optimisation is not true for every AVR board in the repository.

Reply:
After I rebased onto develop, the sizes became a tad larger. No idea why, though.

Reply:
Now with a clean install of QMK MSYS and `avr-gcc 8.5.0`:

With this PR:

```
[andre@Andre-PC qmk_firmware]$ git checkout 32b041900af5f0405bd32e382a89504c7d201a0b && make clean && make git-submodule
[andre@Andre-PC qmk_firmware]$ qmk compile -c -kb cradio -km default -e LTO_ENABLE=yes
 * The firmware size is fine - 16814/28672 (58%, 11858 bytes free)
 ```
 
 Without this PR:
 
 ```
[andre@Andre-PC qmk_firmware]$ git checkout 2e679450336b6679a7cfa4b024d33a10565b2e6 && make clean && make git-submodule
[andre@Andre-PC qmk_firmware]$ qmk compile -c -kb cradio -km default -e LTO_ENABLE=yes
* The firmware size is fine - 16766/28672 (58%, 11906 bytes free)
 ```
 
So yes, there's some size increase involved in the PR as a whole.
 
But, replacing `sym_defer_g.c` with the following contents, I got `16774 bytes`, which is a lot smaller. It seems that the call to the inlined function requires more space than the extra variable.

```c
#include "debounce.h"
#include "timer.h"
#include <string.h>
#ifndef DEBOUNCE
#    define DEBOUNCE 5
#endif

// Maximum debounce: 255ms
#if DEBOUNCE > UINT8_MAX
#    undef DEBOUNCE
#    define DEBOUNCE UINT8_MAX
#endif

#if DEBOUNCE > 0
static bool         debouncing = false;
static fast_timer_t debouncing_time;

#define NOW() (now > 0 ? now : timer_read_fast())

void debounce_init(uint8_t num_rows) {}

bool debounce(matrix_row_t raw[], matrix_row_t cooked[], uint8_t num_rows, bool changed) {
    bool cooked_changed = false;
    fast_timer_t now = 0;

    if (changed) {
        debouncing      = true;
        debouncing_time = now = timer_read_fast();
    }

    if (debouncing && TIMER_DIFF_FAST(NOW(), debouncing_time) >= DEBOUNCE) {
        const size_t matrix_size = num_rows * sizeof(matrix_row_t);
        if (memcmp(cooked, raw, matrix_size) != 0) {
            memcpy(cooked, raw, matrix_size);
            cooked_changed = true;
        }
        debouncing = false;
    }

    return cooked_changed;
}

void debounce_free(void) {}
#else // no debouncing.
#    include "none.c"
#endif
```

Reply:
> But with [25bba8d](https://github.com/qmk/qmk_firmware/commit/25bba8d2807b28518ef251dcb57180d984fcd2e3) I got `16774 bytes`, which is a lot smaller.

I think it is because the `return false` statement is not within that `(changed)` block.

Reply:
I need to check exactly why. Using `changed` again, at all (like `&& !changed` or `else if`, etc.) has some weird side effect and the resulting image grows again, whereas storing `now` and selecting what to use based only on it doesn't result in increased size.

Reply:
I also need to code the test case to prevent this regressing in the future

Reply:
I marked this as resolved. If needed, let's start a new thread.